changefeedccl: add test-only assertion for checkpoint fields invariant

andyyang890 opened the pull request:
Fixes #149100

---

**roachtest: surface changefeed failures in mixed-version CDC tests** 

Previously, if the changefeed failed in a mixed-version CDC test,
the test wouldn't fail until it hit the timeout. We now periodically
check for changefeed failures and fail the test if one is found.

Release note: None

---

**mixedversion: add RunE method to Test** 

This patch adds a `RunE` method to `Test`. Unlike the existing `Run` function,
it will not immediately fatal the test if an error is encountered and will
instead return the error along with the test plan so that callers may decide
how to handle the error.

Release note: None

---

**changefeedccl: add test-only assertion for checkpoint fields invariant** 

This patch adds a test-only assertion that we won't ever see both
checkpoint fields set on a changefeed job progress struct. It is
test-only because a bug that existed on earlier versions of 25.2
could cause both checkpoint fields to be set and so the production
code continues to discard the legacy checkpoint when that happens.

Release note: None

andyyang890 committed:
changefeedccl: add test-only assertion for checkpoint fields invariant

This patch adds a test-only assertion that we won't ever see both
checkpoint fields set on a changefeed job progress struct. It is
test-only because a bug that existed on earlier versions of 25.2
could cause both checkpoint fields to be set and so the production
code continues to discard the legacy checkpoint when that happens.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/149094)
<!-- Reviewable:end -->

andyyang890 committed:
roachtest: surface changefeed failures in mixed-version CDC tests

Previously, if the changefeed failed in a mixed-version CDC test,
the test wouldn't fail until it hit the timeout. We now periodically
check for changefeed failures and fail the test if one is found.

Release note: None

andyyang890 commented:
TeamCity run with 25 runs of `cdc/mixed-version/checkpointing`: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20069775?buildTab=overview&hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true

It shows that the skipping logic works and also that the test still passes most of the time.

DarrylWong commented on pkg/cmd/roachtest/tests/mixed_version_cdc.go:
```diff
@@ -653,12 +672,28 @@ func runCDCMixedVersionCheckpointing(ctx context.Context, t test.Test, c cluster
 
+	// Check before we upgrade to each version whether we're upgrading to
+	// a version that is affeced by the checkpoint fields bug (#148620).
+	mvt.BeforeUpgrade("check if affected by #148620", checkIfAffectedBy148620)
```
I think you could just call this before every invocation of `waitAndValidate` and avoid the need for the `BeforeUpgrade` hook all together?

DarrylWong replied:
Alternatively, maybe we should add a `mvt.RunE` variant which returns an error and the `plan`. Then you can retroactively check `plan.Versions()` to decide whether or not to skip.

andyyang890 replied:
> Alternatively, maybe we should add a `mvt.RunE` variant which returns an error and the `plan`. Then you can retroactively check `plan.Versions()` to decide whether or not to skip.

I went with this idea

andyyang890 committed:
mixedversion: add RunE method to Test

This patch adds a `RunE` method to `Test`. Unlike the existing `Run` function,
it will not immediately fatal the test if an error is encountered and will
instead return the error along with the test plan so that callers may decide
how to handle the error.

Release note: None

DarrylWong reviewed (approved):
mixed version changes LGTM

andyyang890 commented:
TeamCity run with 25 runs of `cdc/mixed-version/checkpointing`:
https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20074546?buildTab=overview&hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true

It shows that most runs still pass, but also both failure modes due to the known checkpoint fields bug are caught and cause the test to be skipped instead of fail:
```
20:58:48   --- SKIP: cdc/mixed-version/checkpointing (N/A)
20:58:48     expected error due to #148620: background step `task-5` returned error: error in task function task-5: changefeed failed: both legacy and current checkpoint set on changefeed job progress
```
```
21:09:36   --- SKIP: cdc/mixed-version/checkpointing (N/A)
21:09:36     expected error due to #148620: background step `task-6` returned error: error in task function task-6: changefeed failed: both legacy and current checkpoint set on change aggregator spec
```
